homeassistant/components/emulated_hue/upnp.py

```diff
@@ -150,7 +151,12 @@ async def async_create_upnp_datagram_endpoint(
     ssdp_socket.setblocking(False)
 
     # Required for receiving multicast
+    # Note: some code duplication from async_upnp_client/ssdp.py here.
     ssdp_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+    with suppress(AttributeError):
+        ssdp_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
```

Comment:
Do we really want this on linux?

https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ
```
Additionally the kernel performs some "special magic" for SO_REUSEPORT sockets that isn't found in other operating systems: For UDP sockets, it tries to distribute datagrams evenly, for TCP listening sockets, it tries to distribute incoming connect requests (those accepted by calling accept()) evenly across all the sockets that share the same address and port combination. Thus an application can easily open the same port in multiple child processes and then use SO_REUSEPORT to get a very inexpensive load balancing.
```

Reply:
@bdraco It's already being used by virtue of consuming async-upnp-client (found in ssdp.py), so it seems probably harmless to use it on Linux. I'd expect there to already be problems if in fact it was going to cause any.

Reply:
FWIW, [this stackoverflow answer](https://stackoverflow.com/a/18450029) suggests that where Multicast is in use, multiple sockets bound with SO_REUSEPORT will all get copies of the data (vs some kind of balancing), which sounds like what one would want.  Still reading around some more just to be certain.

Reply:
@bdraco Just to follow up on this, I wrote a little silly script just to verify that on Linux, each bound multicast socket DOES get a copy of the multicast data when SO_REUSEPORT is set, and confirmed that it returns the expected and same result as FreeBSD in this scenario.  It's small enough to just include it here.  You can prove it to yourself by fixing the HOST_IP value to point to your host machine's address.  You may need to open holes in your firewall to ensure traffic can reach you on the desired port.

On both my FreeBSD (12.x) and Linux (5.x) systems, I get the desired "Success! Every bound socket received the message!" result.

```
import socket
import time

BROADCAST_PORT = 19001
BROADCAST_ADDR = "239.255.255.250"
HOST_IP = '192.168.1.2'


def create_socket() -> socket.socket:
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.setblocking(False)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

    try:
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
    except AttributeError as e:
        print('Unable to set SO_REUSEPORT. This test will not be valid.')
        raise e

    sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
    sock.setsockopt(socket.SOL_IP, socket.IP_MULTICAST_IF, socket.inet_aton(HOST_IP))
    sock.setsockopt(socket.SOL_IP, socket.IP_ADD_MEMBERSHIP, socket.inet_aton(BROADCAST_ADDR) +
                    socket.inet_aton(HOST_IP))
    sock.bind(('', BROADCAST_PORT))

    return sock

if __name__ == '__main__':
    # create three sockets bound to the same address and port, listening to the same multicast address and port
    sockets = []
    for i in range(3):
        sock = create_socket()
        sockets.append(sock)

    # send some data to the multicast address and port, which we want to be received by all our listeners
    sender = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sender.sendto('spatula!'.encode('utf-8'), (BROADCAST_ADDR, BROADCAST_PORT))
    time.sleep(0.250)  # Hackish way of ensuring travel time... this whole script is a hack really.
    sender.close()

    # now try to receive the message that was just sent to the multicast address, closing the sockets after
    # receipt for cleanliness.
    received_count = 0
    for sock in sockets:
        data, sender_addr = sock.recvfrom(1024)
        sock.setsockopt(socket.SOL_IP, socket.IP_DROP_MEMBERSHIP, socket.inet_aton(BROADCAST_ADDR) +
                        socket.inet_aton(HOST_IP))
        sock.close()
        if data:
            received_count += 1

    # moment of truth
    if len(sockets) == received_count:
        print('Success! Every bound socket received the message!')
    else:
        print(f'Failure - received only {received_count} messages instead of {len(sockets)}')
```

Reply:
I'm ok with merging this as-is but it needs to go through beta and not in a patch release so we have a chance to revert it before it reaches production in the event there are unexpected side effects.

Reply:
@bdraco sounds totally reasonable to me, especially where sockets are concerned; as we both well know, it can be tricky-to-impossible to know what every variant of every system is going to do.  Please feel free to reach out any time if any further input from me is needed.